client/verta/verta/tracking/entities/_deployable_entity.py

```diff
@@ -88,6 +90,74 @@ def _get_artifact_msg(self, key):
     def _get_artifact(self, key):
         raise NotImplementedError
 
+    def set_schema(self, input: dict, output: dict = None) -> None:
+        """
+        Sets the input and output schemas of this Deployable Entity. Schemas are stored as
+        model artifacts.
+
+        The output schema is optional.
+
+        To validate a prediction's input and output against these schemas, use the
+        :func:`~verta.registry.validate_schema` decorator.
+
+        Parameters
+        ----------
+        input : dict
+            Input schema as an OpenAPI-compatible json dict. Easiest to create using pydantic.BaseModel.schema() [#]_.
+        output : dict
+            Output schema as an OpenAPI-compatible json dict. Easiest to create using pydantic.BaseModel.schema().
+
+
+        References
+        ----------
+        .. [#] https://docs.pydantic.dev/1.10/usage/schema/
+
+        """
+        if not isinstance(input, dict):
+            raise TypeError(
+                f"`input` must be of type dict, not {type(input)}; did you remember to call `.schema()`?"
+            )
+        if output is not None and not isinstance(output, dict):
+            raise TypeError(
+                f"`output` must be of type dict, not {type(output)}; did you remember to call `.schema()`?"
+            )
+
+        schema = {
+            "input": input,
+        }
+        if output is not None:
+            schema["output"] = output
+
+        # write a temp file because otherwise `log_artifact` will think the artifact contents are the file path
+        with tempfile.NamedTemporaryFile(mode="w") as temp_file:
+            temp_filename = temp_file.name
+            with open(temp_filename, "w") as file:
+                json.dump(schema, file)
```

Comment:
Hmmmmm I might be wrong, but I feel like having a file double-opened in write mode like this risks encountering a deadlock or something. Could we just use the one file handle object?

```suggestion
        with tempfile.NamedTemporaryFile(mode="w+") as temp_file:
            json.dump(schema, temp_file)
            temp_file.flush()  # flush object buffer
            os.fsync(temp_file.fileno())  # flush OS buffer
            temp_file.seek(0)

```

The client uses this approach in a few other places like here: https://github.com/VertaAI/modeldb/blob/7ef3f21e1a6e163a0375e690d0552811bc1927f1/client/verta/tests/versioning/environment/conftest.py#L14-L18

Reply:
And not that it matters today, but opening the file a second time like that wouldn't work on Windows 😆 

> [Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows NT or later).](https://docs.python.org/3.8/library/tempfile.html#tempfile.NamedTemporaryFile)

Reply:
Not that this applies to our customer base necessarily, but Windows is still the most widely used OS globally, [by a long shot](https://www.statista.com/statistics/268237/global-market-share-held-by-operating-systems-since-2009/)

Reply:
It doesn't matter for us today because our client doesn't work in Windows anyway, due to issues elsewhere in the codebase 😅

Reply:
I literally copied this code from the internet and didn't know it was opening a file twice 😅